src/Chapter3.hs

```diff
@@ -1006,6 +1119,132 @@ Implement data types and typeclasses, describing such a battle between two
+  buff _ RunAction = NotAlive
+  buff f _ = Alive f
+
+data Monster a = Monster
+    { monsterHealth :: Health
+    , monsterAttack :: Attack
+    , monsterActions :: [a]
+    } deriving (Show, Eq)
+
+data Knight a = Knight
+    { knightHealth :: Health
+    , knightAttack :: Attack
+    , knightDefense :: Defense
+    , knightActions :: [a]
+    } deriving (Show, Eq)
+
+class Fighter f where
+  -- ehm, not sure how I did it - just tried different approaches
+  -- since actions :: (Action a) => f -> [a] won't let me create instaces
+  -- I believe the problem is the same https://stackoverflow.com/questions/44243367/rigid-type-variable-in-haskell
+  -- at this point I can take it as a belief since can't explain it
+  -- but now I think I can have as many fighters as I probably can
+  -- the only problem is that I can't restrict e.g. that Knight can only have KnightAction actions
+  actions :: (Action a) => f a -> [a]
+
+  increaseHealth :: f a -> IncreaseHealth -> f a
+  increaseHealth f _ = f
+
+  increaseDefense :: f a -> IncreaseDefense -> f a
+  increaseDefense f _ = f
+
+  getAttack :: f a -> Attack
+  receiveAttack :: f a -> Attack -> ActionResult (f a)
+  alive :: f a -> Bool
+
+instance Fighter Monster where
+  actions = monsterActions
+  receiveAttack m (Attack attack) = if alive afterDamage then Alive afterDamage else NotAlive
+      where
+        health = unHealth . monsterHealth $ m
+        afterDamage = m { monsterHealth = Health (health - attack) }
+  getAttack = monsterAttack
+  alive m = (unHealth . monsterHealth $ m) > 0
+
+instance Fighter Knight where
+  actions = knightActions
+  increaseHealth k (IncreaseHealth h) = k { knightHealth = knightHealth k `append` h }
+  increaseDefense k (IncreaseDefense d) = k { knightDefense = knightDefense k `append` d }
+  receiveAttack k (Attack a) = if alive afterDamage then Alive afterDamage else NotAlive
+      where
+        newAttack = a - (unDefense . knightDefense $ k)
+        health = unHealth . knightHealth $ k
+        newHealth = if newAttack < 0 then health else health - newAttack
+        afterDamage = k { knightHealth = Health newHealth }
+  getAttack = knightAttack
+  alive k = (unHealth . knightHealth $ k) > 0
+
+data FightResult a b = FirstWinner a | SecondWinner b deriving (Show, Eq)
+
+data Turn = First | Second
+
+letsFight :: (Fighter f, Action a, Fighter f', Action a') => f a -> f' a' -> FightResult (f a) (f' a')
+letsFight f f' = go (f, cycle (actions f)) (f', cycle (actions f')) First
+  where
+    -- the initial idea was to have one list of all actions taken from first and second fighter one by one
+    -- but I failed to achieve that unless all the actions are of the same type
+    -- then I wanted to just swap second and first on each iteration but still can't achieve that since
+    -- they are of different types though they have the same type class
+    -- and that's why I have this ugly duplication here and the Turn flag which is not very functional I believe
```

Comment:
I believe possible solutions to this task can solve this problem. But I agree with you that ideally, you want somehow to take two different fighters and swap on each iteration. I agree that the current solution involves some code duplication, which is undesirable, but that's still a valid solution!